src/cookie.rs

```diff
@@ -105,12 +106,14 @@ impl<'a> fmt::Debug for Cookie<'a> {
     }
 }
 
+#[cfg(not(target_arch = "wasm32"))]
 pub(crate) fn extract_response_cookie_headers<'a>(
     headers: &'a hyper::HeaderMap,
 ) -> impl Iterator<Item = &'a HeaderValue> + 'a {
     headers.get_all(SET_COOKIE).iter()
 }
 
+#[cfg(not(target_arch = "wasm32"))]
```

Comment:
I might be wrong, but I suspect _this_ is the actual issue that causes this feature not to work as expected.

That is, the current PR adds support for adding cookies to **requests** that are already present in the cookie jar. However, there's no support yet to actually fill the cookie jar with cookies from the previous **response** by looking at the `set-cookie` header.

If we look at the non-wasm implementation here: https://github.com/seanmonstar/reqwest/blob/e9ba0a9dc79f63c3655f334df23b50b9a841e326/src/async_impl/client.rs#L1891-L1900

This is the logic that we need to implement for Wasm, which, from cursory reading, would need to happen in the `fetch` method in this module.

Does  that make sense?

Reply:
I did a bit more digging and tried to implement this myself quickly. I got relatively far, until I got panics because `std::time` isn't available on `wasm32-unknown-unknown`.

The offending code starts at line `1897` above (`set_cookies`), which then goes into the `cookie-store` crate, which goes to the `cookies` crate, which then ends up here:

https://github.com/pfernie/cookie_store/blob/579b91247a4096c69a3bb94abd86de5bd6295bef/src/cookie.rs#L199-L207

Which then calls into  `Cookie::max_age`, which returns `std::time::Duration`, which panics on wasm:

https://github.com/SergioBenitez/cookie-rs/blob/ba60a065ef18af3949966369e4448653c9d25a60/src/lib.rs#L462-L478

This is where I stopped digging, for now, as I ran out of time. There are ways to resolve this (using `js-sys::Date`, et al.) but I haven't dug into that yet, and it does seem like a fairly large chunk of work to get all the plumbing to work together here, potentially having to update external crates, or re-implementing their logic internally for the wasm target.

For anyone interested in picking this up, here's where I left off, based on the existing changes in this PR:

https://github.com/JeanMertz/reqwest/commit/7d238da4a274a628ad8df9f20f07b2ac2752f587

Reply:
Thanks for the detailed summary of your investigation; as you point out the rabbit hole goes down a few levels on this one. I have opened [an issue](https://github.com/pfernie/cookie_store/issues/26) on `cookie_store` to investigate this, although I do not have time to look at this immediately.

Reply:
Incidentally, I just stumbled across https://github.com/time-rs/time/pull/491, which _might_ resolve all of these issues 🪄 _magically_ 🪄, if we wait for that PR to land, and update both `cookie-store` and `cookie-rs` to use the latest version of `time`.

I might pull that PR and all relevant crates locally, and see what happens if I run my project with that change included.

Reply:
The good news is that with that PR in place, things _seem_ to work. There are no panics, and the cookie store does get filled with data.

The potentially bad news is that my wasm test still doesn't pass, whereas the non-wasm one does, and they both do the same number of requests. I can see that the cookie store is way less full on the wasm test, so I still suspect that it might not be storing all `set-cookie` headers somehow, but this might also be something related to my code.

Regardless, no more panics, and the cookie jar is filled with `set-header` responses.

Reply:
So, I figured out the difference between the two platforms.

In the case of the non-wasm target, I get multiple `set-cookie` headers back, each are individually added to the cookie store. On the wasm targets, those multiple headers are merged into a single `set-cookie` header, which trips up the cookie store.

After doing down _another_ rabbit hole, I ended up here: https://fetch.spec.whatwg.org/#headers-class.

The standard specifically calls out this erroneous behaviour of the Fetch API:

> Unlike a [header list](https://fetch.spec.whatwg.org/#concept-header-list), a [Headers](https://fetch.spec.whatwg.org/#headers) object cannot represent more than one `Set-Cookie` [header](https://fetch.spec.whatwg.org/#concept-header). In a way this is problematic as unlike all other headers `Set-Cookie` headers cannot be combined, but since `Set-Cookie` headers are not exposed to client-side JavaScript this is deemed an acceptable compromise. Implementations could choose the more efficient [Headers](https://fetch.spec.whatwg.org/#headers) object representation even for a [header list](https://fetch.spec.whatwg.org/#concept-header-list), as long as they also support an associated data structure for `Set-Cookie` headers.

Now, that call-out does mention that the concept of a "[header list](https://fetch.spec.whatwg.org/#concept-header-list)" does allow multiple `Set-Cookie` headers, but I'm not quite sure if this is exposed in any way. One way would be to get the actual HTTP message body, and parse the relevant headers from there.

I'll see what's possible here. I'm uncertain if we want to fix this in this PR, a follow-up, or not at all (and consider this "working as intended").

Reply:
FYI, I pushed a [second commit](https://github.com/JeanMertz/reqwest/commit/cfa0ac14e04ee03830c419e1bf57b8fd5752b974) to my branch, that resolves the above issues for me. It should be said that the implementation is specific to my needs (i.e. it uses `node-fetch`'s custom [`raw()` method on `Headers`](https://github.com/node-fetch/node-fetch#extract-set-cookie-header)).

This isn't something that can be supported in this library as-is, unless we add some way to optionally access raw cookies that result in `None` if the JS runtime doesn't support this feature (e.g. on the web, when using the native fetch API).

Given all of this, and from what I've come to understand, I assume that for this PR to move forward, we need to wait for the latest release on `time`, and then (some of) the changes I made in https://github.com/JeanMertz/reqwest/commit/7d238da4a274a628ad8df9f20f07b2ac2752f587.

The lack of support for multiple `Set-Header` cookies could be a separate issue, or "working as intended" (sadly).

Reply:
I have tried [48ba15ad56ff8a9e5d65e89d0e3d09abbe2bda37](https://github.com/JeanMertz/reqwest/commit/cfa0ac14e04ee03830c419e1bf57b8fd5752b974). `cargo run` & `cargo build`  run flawlessly and produce the desired results. `cargo clippy` and `cargo test` throw `no method named fetch_credentials_include found for struct reqwest::RequestBuilder in the current scope`. 

Thank you @JeanMertz for linking your work. It has helped a lot.